controllers/resources/lease.go

```diff
@@ -0,0 +1,263 @@
+	maxTimesToExtendLease = 2
+)
+
+type LeaseManager interface {
+	// ObtainNodeLease will attempt to get a node lease with the correct duration, the duration is affected by whether escalation is used and the remediation timeOut.
+	//The first return value (bool) is an indicator whether the lease was obtained, and the second return value (*time.Duration) is an indicator on when a new reconcile should be scheduled (mainly in order to extend the lease)
+	ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error)
+	//UpdateReconcileResults extends leases which needs to be extended and release those that needs to be releases, it is called at the end of reconcile and returns an updated  timeout and error to be returned by reconcile method (which are used for the next iteration of lease management)
+	UpdateReconcileResults(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck, initialRequeue time.Duration, initialErr error) (time.Duration, error)
+}
+
+type nhcLeaseManager struct {
+	client             client.Client
+	commonLeaseManager lease.Manager
+	log                logr.Logger
+}
+
+func NewLeaseManager(client client.Client, log logr.Logger) (LeaseManager, error) {
+	newManager, err := lease.NewManager(client, holderIdentity)
+	if err != nil {
+		log.Error(err, "couldn't initialize lease manager")
+		return nil, err
+	}
+	return &nhcLeaseManager{
+		client:             client,
+		commonLeaseManager: newManager,
+		log:                log.WithName("nhc lease manager"),
+	}, nil
+}
+
+func (m *nhcLeaseManager) ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error) {
+	nodeName := remediationCR.GetName()
+	leaseDuration := m.calculateLeaseDurationBySingleRemediation(remediationCR, nhc)
+	leaseDurationWithBuffer := leaseDuration + LeaseBuffer
+
+	node := &v1.Node{}
+	if err := m.client.Get(context.Background(), types.NamespacedName{Name: nodeName}, node); err != nil {
+		m.log.Error(err, "couldn't obtain node lease node error getting node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	if err := m.commonLeaseManager.RequestLease(context.Background(), node, leaseDurationWithBuffer); err != nil {
+		if _, ok := err.(*lease.AlreadyHeldError); ok {
+			m.log.Info("can't acquire node lease, it is already owned by another owner", "already held error", err)
+			return false, &RequeueIfLeaseTaken, err
+		}
+
+		m.log.Error(err, "couldn't obtain lease for node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	//all good lease created with wanted duration
+	return true, &leaseDuration, nil
+
+}
+
+func (m *nhcLeaseManager) UpdateReconcileResults(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck, initialRequeue time.Duration, initialErr error) (time.Duration, error) {
```

Comment:
This method is giving me headache. 
- we get a list of all nodes, which can be a potentially huge list
- then for each node, we try to get the lease
- all this while we most probably only need handle very few unhealthy node

Do we really need this?
Can't we do everything we need when iterating the (un)healhty nodes in the reconciler instead of the deferred update? E.g. `healthy node -> delete CR -> delete lease`. Or `unhealthy node -> CR already exists? -> check/extend lease duration`... :thinking:

Reply:
The rational was 
 - focus the lease logic in one place instead of spread it over the already crouded Reconcile code
 - "Better safe than sorry" approach which covers all possible leases.
However I do think you have a valid point regarding resource consumption - so no objection trying the alternative as well